### PR TITLE
refactor: add atexit internal wrapper with unregister

### PR DESCRIPTION
**Pull request by jd**
- Python 2 does not have atexit.unregister, so this adds support for it by
  using an internal wrapper.

- The Tracer would never unregister its shutdown function which introduces a
  memory leak. If you had a million tracer over the lifespan of your program,
  then you would have a million tracers kept around to be shutdown at exit
  time.

True story!

### CODE REVIEW

File: ddtrace/internal/atexit.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# -*- encoding: utf-8 -*-
+"""
+An API to provide atexit functionnalities
```
Comment (P403n1x87) [inline ddtrace/internal/atexit.py:3]:
```suggestion
An API to provide atexit functionalities
```

File: tests/conftest.py [hunk 2]
```diff
@@ -14,3 +17,24 @@ def test_spans(tracer):
     container = TracerSpanContainer(tracer)
     yield container
     container.reset()
+
+
+def _call_program(*args):
```
Comment (P403n1x87) [inline tests/conftest.py:22]: A similar helper is already defined in `tests\utils.py`

File: ddtrace/profiling/profiler.py [hunk 3]
```diff
@@ -276,9 +276,7 @@ def stop(self, flush=True):
         for col in reversed(self._collectors):
             col.join()
 
-        # Python 2 does not have unregister
-        if hasattr(atexit, "unregister"):
-            # You can unregister a method that was not registered, so no need to do any other check
-            atexit.unregister(self.stop)
+        # You can unregister a method that was not registered, so no need to do any other check
```
Comment (Kyle-Verhoog) [inline ddtrace/profiling/profiler.py:279]: This comment seems like it should be in the doc-string for `unregister`

File: ddtrace/internal/atexit.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            func(*args, **kwargs)
+
+        _registry.append(_w)
+        return func
+
+    def unregister(func):
+        # type: (typing.Callable[..., None]) -> None
+        """
+        Unregister an exit function which was previously registered using
+        atexit.register
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/atexit.py:54]: Worth pointing out the behaviour when a function that isn't registered is unregistered?

**Review by Kyle-Verhoog — COMMENTED** [review]
:+1: just one nit

File: ddtrace/internal/atexit.py (hunk 4 above)
Comment (jd) [reply ddtrace/internal/atexit.py:54]: For the anecdote, I copied Python's `atexit` docstring.

File: ddtrace/internal/atexit.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            except Exception:
+                # Mimic the behaviour of Python's atexit hooks.
+                log.exception("Error in atexit hook %r", hook)
+
+    atexit.register(_ddtrace_atexit)
+
+    def register(
+        func,  # type: typing.Callable[..., typing.Any]
+        *args,  # type: typing.Any
+        **kwargs,  # type: typing.Any
```
Comment (P403n1x87) [inline ddtrace/internal/atexit.py:39]: Somehow this seems to be invalid syntax on Python 2.7 and 3.5 🤷

Comment (jd) [reply ddtrace/internal/atexit.py:39]: 😢

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**